pyfr/integrators/dual/pseudo/pseudocontrollers.py

```diff
@@ -74,18 +74,127 @@ def conv_mon(self, i, minniters):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        sect = 'solver-time-integrator'
+
+        # Error norm
+        self._norm = self.cfg.get(sect, 'errest-norm', 'l2')
+        if self._norm not in {'l2', 'uniform'}:
+            raise ValueError('Invalid error norm')
+
+        tplargs = {'ndims': self.system.ndims, 'nvars': self.system.nvars}
+
+        # Error tolerance
+        tplargs['atol'] = self.cfg.getfloat(sect, 'atol')
+
+        # PI control values
+        sord = self._pseudo_stepper_order
+        tplargs['expa'] = self.cfg.getfloat(sect, 'pi-alpha', 0.7) / sord
+        tplargs['expb'] = self.cfg.getfloat(sect, 'pi-beta', 0.4) / sord
+
+        # Constants
+        tplargs['maxf'] = self.cfg.getfloat(sect, 'max-fact', 1.01)
+        tplargs['minf'] = self.cfg.getfloat(sect, 'min-fact', 0.98)
+        tplargs['saff'] = self.cfg.getfloat(sect, 'safety-fact', 0.8)
+        tplargs['dtau_maxf'] = self.cfg.getfloat(sect, 'dtau-max-fact', 3.0)
+
+        # Limits for the local pseudo-time-step size
+        tplargs['dtau_min'] = self._dtau
+        tplargs['dtau_max'] = tplargs['dtau_maxf'] * self._dtau
+
+        # Register a kernel to compute local error
+        self.backend.pointwise.register(
+            'pyfr.integrators.dual.pseudo.kernels.localerrest'
+        )
+
+        for ele, shape, dtau in zip(self.system.ele_map.values(),
+                                    self.system.ele_shapes, self.dtau_upts):
+            # Allocate storage for previous error
+            err_prev = self.backend.matrix(shape,
+                                           np.ones(shape), tags={'align'})
+
+            # Append the error kernels to the proxylist
+            self.pintgkernels['eles', 'localerrest'].append(
+                self.backend.kernel(
+                    'localerrest', tplargs=tplargs,
+                    dims=[ele.nupts, ele.neles], err=ele.scal_upts_inb,
+                    errprev=err_prev, dtau_upts=dtau
+                )
+            )
+
+        self.backend.commit()
+
+    @property
+    def _pseudo_controller_needs_lerrest(self):
+        return True
+
+    def localerrest(self, errbank):
+        self.system.eles_scal_upts_inb.active = errbank
+        self._queue % self.pintgkernels['eles', 'localerrest']()
+
+    def convmon(self, i, minniters, dtau=None):
+        if i >= minniters - 1:
+            # Subtract the current solution from the previous solution
+            self._add(-1.0, self._idxprev, 1.0, self._idxcurr)
+
+            # Divide by 1/dtau
+            self.localdtau(self._idxprev, inv=1)
+
+            # Reduction
+            resid = tuple(self._resid(1.0, self._idxprev))
+        else:
+            resid = None
+
+        # Increment the step count
+        self.npseudosteps += 1
+
+        self.pseudostepinfo.append((self.npseudosteps, i + 1, resid))
+
+        converged = all(r <= t for r, t in zip(resid, self._pseudo_residtol))
+
```

Comment:
Remove newline.

Reply:
Actually, this does not work with resid=None. I'm about to push changes.